Zend/zend_vm_def.h

```diff
@@ -6373,6 +6373,60 @@ ZEND_VM_C_LABEL(isset_no_object):
 	ZEND_VM_NEXT_OPCODE_CHECK_EXCEPTION();
 }
 
+ZEND_VM_HANDLER(199, ZEND_ARRAY_KEY_EXISTS, CV|TMPVAR|CONST, CV|TMPVAR|CONST)
+{
+	USE_OPLINE
+
+	zend_free_op free_op1, free_op2;
+	zval *subject, *key;
+	int result;
+
+	SAVE_OPLINE();
+
+	subject = GET_OP2_ZVAL_PTR(BP_VAR_R);
+	key = GET_OP1_ZVAL_PTR(BP_VAR_R);
+
+	if (UNEXPECTED(Z_ISREF_P(subject))) {
+		subject = Z_REFVAL_P(subject);
+	}
+	if (UNEXPECTED(Z_ISREF_P(key))) {
+		key = Z_REFVAL_P(key);
+	}
+
+	if (EXPECTED(Z_TYPE_P(subject) == IS_ARRAY)) {
+		if (EXPECTED(Z_TYPE_P(key) == IS_STRING)) {
+			result = zend_symtable_exists_ind(Z_ARRVAL_P(subject), Z_STR_P(key));
+		} else if (EXPECTED(Z_TYPE_P(key) == IS_LONG)) {
+			result = zend_hash_index_exists(Z_ARRVAL_P(subject), Z_LVAL_P(key));
+		} else if (UNEXPECTED(Z_TYPE_P(key) == IS_NULL)) {
+			result = zend_symtable_exists_ind(Z_ARRVAL_P(subject), ZSTR_EMPTY_ALLOC());
+		} else {
+			zend_error(E_WARNING, "array_key_exists(): The first argument should be either a string or an integer");
+			result = 0;
+		}
+		ZVAL_BOOL(EX_VAR(opline->result.var), result);
+	} else if (UNEXPECTED(Z_TYPE_P(subject) == IS_OBJECT)) {
+		if (EXPECTED(Z_TYPE_P(key) == IS_STRING)) {
+			result = zend_symtable_exists_ind(Z_OBJPROP_P(subject), Z_STR_P(key));
+		} else if (EXPECTED(Z_TYPE_P(key) == IS_LONG)) {
+			result = zend_hash_index_exists(Z_OBJPROP_P(subject), Z_LVAL_P(key));
+		} else if (UNEXPECTED(Z_TYPE_P(key) == IS_NULL)) {
+			result = zend_symtable_exists_ind(Z_OBJPROP_P(subject), ZSTR_EMPTY_ALLOC());
+		} else {
+			zend_error(E_WARNING, "array_key_exists(): The first argument should be either a string or an integer");
+			result = 0;
+		}
+		ZVAL_BOOL(EX_VAR(opline->result.var), result);
```

Comment:
The code between IS_ARRAY and IS_OBJECT is duplicated here (which is actually rather dubious and will handle some cases incorrectly, but that's an existing issue and best solved by deprecating array_key_exists on objects entirely). It would be better to only extract the hashtable using Z_ARRVAL_P and Z_OBJPROP_P in these branches and have only one copy of the common key handling code afterwards.